pandas/core/arrays/datetimes.py

```diff
@@ -206,45 +206,103 @@ def _simple_new(cls, values, freq=None, tz=None):
 
         freq, freq_infer = dtl.maybe_infer_freq(freq)
+        if freq is None and hasattr(data, "freq"):
+            # i.e. DatetimeArray/Index
+            freq = data.freq
+            verify_integrity = False
 
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
-        if is_object_dtype(values):
-            # kludge; dispatch until the DatetimeArray constructor is complete
-            from pandas import DatetimeIndex
-            values = DatetimeIndex(values, freq=freq, tz=tz)
-
-        if isinstance(values, ABCSeries):
-            # extract to ndarray or DatetimeIndex
-            values = values._values
+        # By this point we are assured to have either a numpy array or Index
+        data, copy = maybe_convert_dtype(data, copy)
 
-        if isinstance(values, DatetimeArrayMixin):
-            # extract nanosecond unix timestamps
-            if tz is None:
-                tz = values.tz
-            values = values.asi8
-
-        if values.dtype == 'i8':
-            values = values.view('M8[ns]')
+        if is_object_dtype(data) or is_string_dtype(data):
+            # TODO: We do not have tests specific to string-dtypes,
+            #  also complex or categorical or other extension
+            copy = False
+            if lib.infer_dtype(data) == 'integer':
+                data = data.astype(np.int64)
+            else:
+                # data comes back here as either i8 to denote UTC timestamps
+                #  or M8[ns] to denote wall times
+                data, inferred_tz = objects_to_datetime64ns(
+                    data, dayfirst=dayfirst, yearfirst=yearfirst)
+                tz = maybe_infer_tz(tz, inferred_tz)
+
+        if is_datetime64tz_dtype(data):
+            tz = maybe_infer_tz(tz, data.tz)
+            subarr = data._data
+
+        elif is_datetime64_dtype(data):
+            # tz-naive DatetimeArray/Index or ndarray[datetime64]
+            data = getattr(data, "_data", data)
+            if data.dtype != _NS_DTYPE:
+                data = conversion.ensure_datetime64ns(data)
+
+            if tz is not None:
+                # Convert tz-naive to UTC
+                tz = timezones.maybe_get_tz(tz)
+                data = conversion.tz_localize_to_utc(data.view('i8'), tz,
+                                                     ambiguous=ambiguous)
+                data = data.view(_NS_DTYPE)
+
+            assert data.dtype == _NS_DTYPE, data.dtype
+            subarr = data
 
-        assert isinstance(values, np.ndarray), type(values)
-        assert is_datetime64_dtype(values)  # not yet assured nanosecond
-        values = conversion.ensure_datetime64ns(values, copy=False)
+        else:
+            # must be integer dtype otherwise
+            # assume this data are epoch timestamps
+            if data.dtype != _INT64_DTYPE:
+                data = data.astype(np.int64, copy=False)
+            subarr = data.view(_NS_DTYPE)
```

Comment:
might be able to pull out
``subarr = data.view(_NS_DTYPE)`` of the if/else and just do it always (or maybe just do it in _simple_new), but this is for later

Reply:
That's what I tried originally, but it breaks the recently-implemented tests.arrays.test_datetimelike.test_from_array_keeps_base (#23956)

Reply:
Hrm sorry. FWIW, that was a hackish thing that I introduced to avoid segfaults. If the code in reduction code was doing the right thing, we may be able to remove that test / requirement. But I'm not sure what the right thing is.

Reply:
For the purposes of this PR it sounds like the options are to either remove that test and un-indent t his line, or keep this line how it is.  It sounds like the first option would cause problems in your DTA branch.  I don't have a strong preference here.